In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/datasets/sudalairajkumar/indian-startup-funding/startup_funding.csv


# **Objective**

The goal of this notebook is to clean and preprocess the Indian Startup Funding dataset for exploratory data analysis and dashboard development.

In [3]:
df= pd.read_csv ('/kaggle/input/datasets/sudalairajkumar/indian-startup-funding/startup_funding.csv')

### Dataset Information

The dataset contains funding details of Indian startups including:
- Startup Name
- Industry Vertical
- Investors
- Funding Amount
- Funding Round
- City Location
- Funding Date

In [4]:
df.head()

,Sr No,Date dd/mm/yyyy,Startup Name,Industry Vertical,SubVertical,City Location,Investors Name,InvestmentnType,Amount in USD,Remarks
0,1,09/01/2020,BYJU’S,E-Tech,E-learning,Bengaluru,Tiger Global Management,Private Equity Round,"20,00,00,000",NaN
1,2,13/01/2020,Shuttl,Transportation,App based shuttle service,Gurgaon,Susquehanna Growth Equity,Series C,"80,48,394",NaN
2,3,09/01/2020,Mamaearth,E-commerce,Retailer of baby and toddler products,Bengaluru,Sequoia Capital India,Series B,"1,83,58,860",NaN
3,4,02/01/2020,https://www.wealthbucket.in/,FinTech,Online Investment,New Delhi,Vinod Khatumal,Pre-series A,"30,00,000",NaN
4,5,02/01/2020,Fashor,Fashion and Apparel,Embroiled Clothes For Women,Mumbai,Sprout Venture Partners,Seed Round,"18,00,000",NaN


## **Inpecting the Data**

In [5]:
# Shape of dataset
print("Rows:", df.shape[0], "Columns:", df.shape[1])

# Info
df.info()

# Missing values
df.isnull().sum()

Rows: 3044 Columns: 10
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3044 entries, 0 to 3043
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Sr No              3044 non-null   int64 
 1   Date dd/mm/yyyy    3044 non-null   object
 2   Startup Name       3044 non-null   object
 3   Industry Vertical  2873 non-null   object
 4   SubVertical        2108 non-null   object
 5   City  Location     2864 non-null   object
 6   Investors Name     3020 non-null   object
 7   InvestmentnType    3040 non-null   object
 8   Amount in USD      2084 non-null   object
 9   Remarks            419 non-null    object
dtypes: int64(1), object(9)
memory usage: 237.9+ KB


Sr No                   0
Date dd/mm/yyyy         0
Startup Name            0
Industry Vertical     171
SubVertical           936
City  Location        180
Investors Name         24
InvestmentnType         4
Amount in USD         960
Remarks              2625
dtype: int64

**Renaming Column Name**

In [6]:
df.rename(columns={
    'Sr No' : 'Sr_No',
    'Date dd/mm/yyyy': 'Date',
    'Startup Name' : 'StartUp',
    'Industry Vertical' : 'Vertical',
    'City  Location' : 'Location',
    'Investors Name' : 'Investors',
    'InvestmentnType' : 'Round',
    'Amount in USD' : 'Amount'
},inplace = True)

In [7]:
df.columns

Index(['Sr_No', 'Date', 'StartUp', 'Vertical', 'SubVertical', 'Location',
       'Investors', 'Round', 'Amount', 'Remarks'],
      dtype='object')

In [8]:
df.set_index('Sr_No')

,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount,Remarks
Sr_No,,,,,,,,,
1,09/01/2020,BYJU’S,E-Tech,E-learning,Bengaluru,Tiger Global Management,Private Equity Round,"20,00,00,000",NaN
2,13/01/2020,Shuttl,Transportation,App based shuttle service,Gurgaon,Susquehanna Growth Equity,Series C,"80,48,394",NaN
3,09/01/2020,Mamaearth,E-commerce,Retailer of baby and toddler products,Bengaluru,Sequoia Capital India,Series B,"1,83,58,860",NaN
4,02/01/2020,https://www.wealthbucket.in/,FinTech,Online Investment,New Delhi,Vinod Khatumal,Pre-series A,"30,00,000",NaN
5,02/01/2020,Fashor,Fashion and Apparel,Embroiled Clothes For Women,Mumbai,Sprout Venture Partners,Seed Round,"18,00,000",NaN
...,...,...,...,...,...,...,...,...,...
3040,29/01/2015,Printvenue,NaN,NaN,NaN,Asia Pacific Internet Group,Private Equity,"45,00,000",NaN
3041,29/01/2015,Graphene,NaN,NaN,NaN,KARSEMVEN Fund,Private Equity,"8,25,000",Govt backed VC Fund
3042,30/01/2015,Mad Street Den,NaN,NaN,NaN,"Exfinity Fund, GrowX Ventures.",Private Equity,"15,00,000",NaN


**Correcting the Data Type of Columns**

In [9]:
df['Date'] = pd.to_datetime(df['Date'], format='%d/%m/%Y', errors='coerce')
df['Amount'] = pd.to_numeric(df['Amount'].str.replace(',', ''), errors='coerce')


In [10]:
#df['Amount'] = df['Amount'].str.replace(',','').astype('int64')

- This ValueError occurs because your Amount column contains the text string 'undisclosed', which cannot be converted directly into an integer.
- To fix this, you should use pd.to_numeric() with the errors='coerce' parameter. This will automatically turn non-numeric text like "undisclosed" into NaN (null) values, allowing the rest of the conversion to proceed.
- With errors = 'coerce' If you run into a value that you can't convert, don't crash; just turn it into a NaN

### **Handling the Missing value**

In [11]:
df.isnull().sum()

Sr_No             0
Date              8
StartUp           0
Vertical        171
SubVertical     936
Location        180
Investors        24
Round             4
Amount          979
Remarks        2625
dtype: int64

- Whole row is null:  > df[df.isnull().all(axis=1)]
- At least one null in the row:  > df[df.isnull().any(axis=1)]

In [12]:
df[df.isnull().all(axis=1)]

,Sr_No,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount,Remarks


#### Fixing the inconsistency in the 'Amount' column -

In [13]:
df['Amount'].head()

0    200000000.0
1      8048394.0
2     18358860.0
3      3000000.0
4      1800000.0
Name: Amount, dtype: float64

**Change USD $ to INR (Rs)**

In [14]:
def to_inr(dollar):
    inr = dollar * 94.43  #Convert USD to INR
    return inr/10000000   #Converting INR to Carores

In [15]:
df['Amount']=df['Amount'].apply(to_inr)

In [16]:
df['Amount'].head()

0    1888.600000
1      76.000985
2     173.362715
3      28.329000
4      16.997400
Name: Amount, dtype: float64

In [17]:
#Filling the missing value in the Amount column 
mean= df['Amount'].mean()
median = df['Amount']. median()

print(f'Mean = {mean}')
print(f'Median = {median}')

Mean = 174.0335199282467
Median = 16.0531


Mean (174.03) is significantly higher than your Median (16.05). This usually means your 'Amount' column has outliers

In [18]:
# Calculate Q1 (25th percentile) and Q3 (75th percentile)
Q1 = df['Amount'].quantile(0.25)
Q3 = df['Amount'].quantile(0.75)
IQR = Q3 - Q1

# Define bounds (standard is 1.5 * IQR)
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter the DataFrame to remove outliers
df_cleaned = df[(df['Amount'] >= lower_bound) & (df['Amount'] <= upper_bound)]

print(f"Old shape: {df.shape}")
print(f"New shape: {df_cleaned.shape}")


Old shape: (3044, 10)
New shape: (1782, 10)


- Row count dropped from 3044 to 1782. This happened because the filter (df['Amount'] >= lower_bound) automatically excludes rows where the amount is NaN.

In [19]:
new_mean = df_cleaned['Amount'].mean()
new_median = df_cleaned['Amount'].median()

print(f'New Mean = {new_mean}')
print(f'New Median = {new_median}')


New Mean = 29.835263762530563
New Median = 9.443


Even after removing outliers, your Mean (29.8) is still about 3 times higher than your Median (9.44). This suggests your data is still "right-skewed

In [20]:
df['Amount'] = df['Amount'].fillna(new_median)

- Using the Median to fill 939 rows (about 30% of your data) is a common practice, but it comes with a trade-off you should consider.
- Filling nearly 1,000 rows with the exact same number (9.44) will create a huge "spike" in your data. This can make your correlations look stronger or weaker than they actually are because you've artificially reduced the variety in the 'Amount' column.

**Outlier Handling Strategy**

* Outliers were detected using the IQR method. Outliers were excluded only during median calculation for missing value imputation to obtain a more stable and representative median value.
* Outliers were not removed from the dataset because startup funding data is naturally highly skewed. Large funding amounts from well-funded startups represent real-world business scenarios and contain important analytical information.

* Removing these values could lead to loss of meaningful insights related to major investments and funding trends.



In [21]:
# Checking the misisng value percentage -
df.isnull().sum()/len(df) *100

Sr_No           0.000000
Date            0.262812
StartUp         0.000000
Vertical        5.617608
SubVertical    30.749014
Location        5.913272
Investors       0.788436
Round           0.131406
Amount          0.000000
Remarks        86.235217
dtype: float64

In [22]:
# Remark has more that 80 perc of missing value --> So we ca remove them 
df = df.drop(columns=['Remarks'])

In [23]:
df[df['Vertical'].isnull()]

,Sr_No,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount
2873,2874,2015-03-02,TrulyMadly.com,NaN,NaN,NaN,"Helion Venture Partners, Kae Capital",Private Equity,51.936500
2874,2875,2015-03-02,InstaLively,NaN,NaN,NaN,Group of Angel Investors,Seed Funding,1.133160
2875,2876,2015-03-03,Vserv,NaN,NaN,NaN,"Maverick Capital, IDG Ventures India",Private Equity,141.645000
2876,2877,2015-03-03,Intruo.com,NaN,NaN,NaN,"Ashutosh Lawania, Alok Goel & others",Seed Funding,9.443000
2877,2878,2015-03-05,Niffler,NaN,NaN,NaN,SAIF Partners,Seed Funding,9.443000
...,...,...,...,...,...,...,...,...,...
3039,3040,2015-01-29,Printvenue,NaN,NaN,NaN,Asia Pacific Internet Group,Private Equity,42.493500
3040,3041,2015-01-29,Graphene,NaN,NaN,NaN,KARSEMVEN Fund,Private Equity,7.790475
3041,3042,2015-01-30,Mad Street Den,NaN,NaN,NaN,"Exfinity Fund, GrowX Ventures.",Private Equity,14.164500
3042,3043,2015-01-30,Simplotel,NaN,NaN,NaN,MakeMyTrip,Private Equity,9.443000


In [24]:
df[df['Location'].isnull()]

,Sr_No,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount
338,339,2018-03-20,One Labs,Consumer Internet,Consumer Technology and Artificial Intelligenc...,NaN,Micromax Informatics Ltd,Private Equity,9.443000
1600,1601,2016-06-13,Matrubharti,Consumer Internet,e-Book Publisher,NaN,Viridian Capital,Seed Funding,0.283290
1765,1766,2016-04-12,Indi.com,Consumer Internet,Video Social Network,NaN,Anil Kapoor,Seed Funding,9.443000
1811,1812,2016-04-27,CareonGO,Consumer Internet,Online Pharmacy,NaN,"Farooq Oomerbhoy, Anupam Mittal, Anand Mittal,...",Seed Funding,9.443000
1891,1892,2016-03-23,HashLearn,Consumer Internet,Mobile based private tutoring services,NaN,"Bhavish Aggarwal, Binny Bansal,\\xc2\\xa0 Raja...",Seed Funding,9.443000
...,...,...,...,...,...,...,...,...,...
3039,3040,2015-01-29,Printvenue,NaN,NaN,NaN,Asia Pacific Internet Group,Private Equity,42.493500
3040,3041,2015-01-29,Graphene,NaN,NaN,NaN,KARSEMVEN Fund,Private Equity,7.790475
3041,3042,2015-01-30,Mad Street Den,NaN,NaN,NaN,"Exfinity Fund, GrowX Ventures.",Private Equity,14.164500
3042,3043,2015-01-30,Simplotel,NaN,NaN,NaN,MakeMyTrip,Private Equity,9.443000


In [25]:
df= df.dropna(subset=['Vertical','SubVertical','Location'], how= 'all')

In [26]:
# Filling the missing value
# 'SubVertical' (fillng it with unknown)
df['SubVertical'] = df['SubVertical'].fillna("Unknown")


In [27]:
#Filling the Investors columns
df['Investors'] = df['Investors'].fillna('Undisclosed')

In [28]:
# Checking the misisng value percentage -
df.isnull().sum()

Sr_No          0
Date           6
StartUp        0
Vertical       0
SubVertical    0
Location       9
Investors      0
Round          4
Amount         0
dtype: int64

In [29]:
df = df.dropna(subset = ['Date', 'StartUp', 'Vertical', 'Location', 'Investors', 'Round','Amount'])#

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2854 entries, 0 to 2872
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Sr_No        2854 non-null   int64         
 1   Date         2854 non-null   datetime64[ns]
 2   StartUp      2854 non-null   object        
 3   Vertical     2854 non-null   object        
 4   SubVertical  2854 non-null   object        
 5   Location     2854 non-null   object        
 6   Investors    2854 non-null   object        
 7   Round        2854 non-null   object        
 8   Amount       2854 non-null   float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 223.0+ KB


In [31]:
#Checking for duplicate rows
duplicates = df.duplicated().sum()
print(f'The Dataframe has {duplicates} no. rows ')

The Dataframe has 0 no. rows 


#### **Fixing the inconsistency in the 'StartUp' column -**

In [32]:
#Getting all the startup name 
df['StartUp'].unique().shape 

(2324,)

In [33]:
df['StartUp'].unique().tolist()

['BYJU’S',
 'Shuttl',
 'Mamaearth',
 'https://www.wealthbucket.in/',
 'Fashor',
 'Pando',
 'Zomato',
 'Ecozen',
 'CarDekho',
 'Dhruva Space',
 'Rivigo',
 'Healthians',
 'Licious',
 'InCred',
 'Trell',
 'Rein Games',
 'Lenskart.com',
 'Freshworks',
 'Misters',
 'Sunstone Eduversity Pvt. Ltd',
 'Burger Singh',
 'Ninjacart',
 'Aye Finance',
 'SuperGaming',
 'Clumio',
 'eBikeGo',
 'Digital Mall Asia',
 'Medikabazaar',
 'Vogo Automotive',
 'Paytm',
 'Dunzo',
 'Udaan',
 'The Man Company',
 'FPL Technologies',
 'Cashflo',
 'Digital F5',
 '3rdFlix',
 '75F',
 'Myelin Foundry',
 'Atomberg Technology',
 'GOQii',
 'Vyapar App',
 'Progcap',
 'MyPetrolPump',
 'Alteria Capital',
 'Pine Labs',
 'Meesho',
 'Cars24',
 'Uniphore',
 'Zendrive',
 'Lo! Foods',
 'Tala',
 'INDwealth',
 'HungerBox',
 'AdmitKard',
 'Mishry Reviews',
 'Grofers',
 'Rapido Bike Taxi',
 'RenewBuy',
 'Atlan',
 'WizCounsel',
 'Ola Cabs',
 'Daalchini Technologies',
 '"BYJU\\\\\'S"',
 'Moglix',
 'Ezyhaul',
 'Indus OS',
 'HealthAssure',

In [34]:
#By manual assessment we get the inconsistency in startup names ----> Resolving that 
df['StartUp'] = df['StartUp'].str.replace(r'\\\\xe2\\\\x80\\\\x99', "'", regex=True).str.replace(r'\\\\xe2\\\\x80\\\\x9d', '"', regex=True).str.replace(r'\\\\x[a-f0-9]{2}', '', regex=True).str.replace('https://www.wealthbucket.in/','Wealth Bucket').str.replace(r'\s+', ' ', regex=True).str.strip()
df['StartUp'] = df['StartUp'].str.title().str.strip()

In [35]:
df['StartUp'].unique().tolist()

['Byju’S',
 'Shuttl',
 'Mamaearth',
 'Wealth Bucket',
 'Fashor',
 'Pando',
 'Zomato',
 'Ecozen',
 'Cardekho',
 'Dhruva Space',
 'Rivigo',
 'Healthians',
 'Licious',
 'Incred',
 'Trell',
 'Rein Games',
 'Lenskart.Com',
 'Freshworks',
 'Misters',
 'Sunstone Eduversity Pvt. Ltd',
 'Burger Singh',
 'Ninjacart',
 'Aye Finance',
 'Supergaming',
 'Clumio',
 'Ebikego',
 'Digital Mall Asia',
 'Medikabazaar',
 'Vogo Automotive',
 'Paytm',
 'Dunzo',
 'Udaan',
 'The Man Company',
 'Fpl Technologies',
 'Cashflo',
 'Digital F5',
 '3Rdflix',
 '75F',
 'Myelin Foundry',
 'Atomberg Technology',
 'Goqii',
 'Vyapar App',
 'Progcap',
 'Mypetrolpump',
 'Alteria Capital',
 'Pine Labs',
 'Meesho',
 'Cars24',
 'Uniphore',
 'Zendrive',
 'Lo! Foods',
 'Tala',
 'Indwealth',
 'Hungerbox',
 'Admitkard',
 'Mishry Reviews',
 'Grofers',
 'Rapido Bike Taxi',
 'Renewbuy',
 'Atlan',
 'Wizcounsel',
 'Ola Cabs',
 'Daalchini Technologies',
 '"Byju\\\\\'S"',
 'Moglix',
 'Ezyhaul',
 'Indus Os',
 'Healthassure',
 'House Of Msa

#### **Fixing the inconsistency in the 'Vertical' column -**

In [36]:
df['Vertical'].unique().tolist()

['E-Tech',
 'Transportation',
 'E-commerce',
 'FinTech',
 'Fashion and Apparel',
 'Logistics',
 'Hospitality',
 'Technology',
 'E-Commerce',
 'Aerospace',
 'B2B-focused foodtech startup',
 'Finance',
 'Video',
 'Gaming',
 'Software',
 'Health and wellness',
 'Education',
 'Food and Beverage',
 'Health and Wellness',
 'B2B Marketing',
 'Video Games',
 'SaaS',
 'Last Mile Transportation',
 'Healthcare',
 'Customer Service',
 'B2B',
 'Consumer Goods',
 'Advertising, Marketing',
 'IoT',
 'Information Technology',
 'Consumer Technology',
 'Accounting',
 'Retail',
 'Customer Service Platform',
 'Automotive',
 'EdTech',
 'Services',
 'Compliance',
 'Transport',
 'Artificial Intelligence',
 'Tech',
 'Health Care',
 'Luxury Label',
 'Waste Management Service',
 'Deep-Tech',
 'Energy',
 'Digital Media',
 'Saas',
 'Automobile',
 'Agtech',
 'Social Media',
 'Fintech',
 'Edtech',
 'AI',
 'Ecommerce',
 'Nanotechnology',
 'Services Platform',
 'Travel Tech',
 'Online Education',
 'Online Marketplace'

In [37]:
#By manual assessment we get the inconsistency in vertical names ----> Resolving that 
#Fix the 'é' (like in Café or Décor)
df['Vertical'] = df['Vertical'].str.replace(r'\\\\xc3\\\\xa9', 'e', regex=True)

# 2. Fix the apostrophe (like in Byju's or Caf's)
df['Vertical'] = df['Vertical'].str.replace(r'\\\\xe2\\\\x80\\\\x99', "'", regex=True)

# 3. Wipe out the broken space junk (\\\\xc2\\\\xa0)
df['Vertical'] = df['Vertical'].str.replace(r'\\\\xc2\\\\xa0', ' ', regex=True)

# 4. Fix Newlines (\\\\Nplatform) found in 'Mobile Growth Hacking'
df['Vertical'] = df['Vertical'].str.replace(r'\\\\n', ' ', regex=True, case=False)

# 5. Final Polish: Fix casing and remove extra whitespace
df['Vertical'] = df['Vertical'].str.strip().str.title()

# 6. Optional: Clean up repetitive suffixes to help grouping
# This merges "Booking Platform" and "Booking App" into just "Booking"
df['Vertical'] = df['Vertical'].str.replace(r'\s+(Platform|App|Marketplace|Service|Solution)s?\b', '', regex=True, case=False)

In [38]:
df['Vertical'].unique().tolist()

['E-Tech',
 'Transportation',
 'E-Commerce',
 'Fintech',
 'Fashion And Apparel',
 'Logistics',
 'Hospitality',
 'Technology',
 'Aerospace',
 'B2B-Focused Foodtech Startup',
 'Finance',
 'Video',
 'Gaming',
 'Software',
 'Health And Wellness',
 'Education',
 'Food And Beverage',
 'B2B Marketing',
 'Video Games',
 'Saas',
 'Last Mile Transportation',
 'Healthcare',
 'Customer',
 'B2B',
 'Consumer Goods',
 'Advertising, Marketing',
 'Iot',
 'Information Technology',
 'Consumer Technology',
 'Accounting',
 'Retail',
 'Automotive',
 'Edtech',
 'Services',
 'Compliance',
 'Transport',
 'Artificial Intelligence',
 'Tech',
 'Health Care',
 'Luxury Label',
 'Waste Management',
 'Deep-Tech',
 'Energy',
 'Digital Media',
 'Automobile',
 'Agtech',
 'Social Media',
 'Ai',
 'Ecommerce',
 'Nanotechnology',
 'Travel Tech',
 'Online Education',
 'Online',
 'Saas, Ecommerce',
 'Nbfc',
 'Food',
 'Food Tech',
 'Automation',
 'Investment',
 'Social Network',
 'Financial Tech',
 'Fashion',
 'Real Estate',
 

#### **Fixing the inconsistency in the 'SubVertical' column -**

In [39]:
df['SubVertical'].unique().tolist()

['E-learning',
 'App based shuttle service',
 'Retailer of baby and toddler products',
 'Online Investment',
 'Embroiled Clothes For Women',
 'Open-market, freight management platform',
 'Online Food Delivery Platform',
 'Agritech',
 'Automobile',
 'Satellite Communication',
 'Logistics Services and Solutions',
 'Food Solutions For Corporate',
 'Online Meat And Seafood Ordering Startup',
 'Non-Banking Financial Company',
 'Experience Discovery Platform',
 'Real money based gaming startup',
 'Online Eyewear Shopping Portal',
 'Business and customer engagement tools',
 "Men's Health and Wellness brand",
 'Elearning',
 'Indian Burger Brand',
 'Healthcare services',
 'Financial Services To MSMEs',
 'Social gaming platform',
 'Recovery software',
 'Electric bike rental',
 'Virtual e-commerce platform',
 'B2B platform for medical supplies',
 'Scooter sharing app',
 'Mobile Wallet',
 'Delivery Service',
 'Business development',
 'Beauty and Grooming',
 'Financial Services',
 'Invoice discount

In [40]:
# 1. Fix the broken apostrophe (\\\\xe2\\\\x80\\\\x99)
df['SubVertical'] = df['SubVertical'].str.replace(r'\\\\xe2\\\\x80\\\\x99', "'", regex=True)

# 2. Fix the broken space junk (\\\\xc2\\\\xa0) from previous examples
df['SubVertical'] = df['SubVertical'].str.replace(r'\\\\xc2\\\\xa0', ' ', regex=True)

# 3. Standardize casing and trim extra whitespace
df['SubVertical'] = df['SubVertical'].str.title().str.strip()

In [41]:
df['SubVertical'].unique().tolist()

['E-Learning',
 'App Based Shuttle Service',
 'Retailer Of Baby And Toddler Products',
 'Online Investment',
 'Embroiled Clothes For Women',
 'Open-Market, Freight Management Platform',
 'Online Food Delivery Platform',
 'Agritech',
 'Automobile',
 'Satellite Communication',
 'Logistics Services And Solutions',
 'Food Solutions For Corporate',
 'Online Meat And Seafood Ordering Startup',
 'Non-Banking Financial Company',
 'Experience Discovery Platform',
 'Real Money Based Gaming Startup',
 'Online Eyewear Shopping Portal',
 'Business And Customer Engagement Tools',
 "Men'S Health And Wellness Brand",
 'Elearning',
 'Indian Burger Brand',
 'Healthcare Services',
 'Financial Services To Msmes',
 'Social Gaming Platform',
 'Recovery Software',
 'Electric Bike Rental',
 'Virtual E-Commerce Platform',
 'B2B Platform For Medical Supplies',
 'Scooter Sharing App',
 'Mobile Wallet',
 'Delivery Service',
 'Business Development',
 'Beauty And Grooming',
 'Financial Services',
 'Invoice Discount

In [42]:
#### **Fixing the inconsistency in the 'Round' column -**

In [43]:
df['Round'].unique().tolist()

['Private Equity Round',
 'Series C',
 'Series B',
 'Pre-series A',
 'Seed Round',
 'Series A',
 'Series D',
 'Seed',
 'Series F',
 'Series E',
 'Debt Funding',
 'Series G',
 'Series H',
 'Venture',
 'Seed Funding',
 'Funding Round',
 'Corporate Round',
 'Maiden Round',
 'pre-series A',
 'Seed Funding Round',
 'Single Venture',
 'Venture Round',
 'Pre-Series A',
 'Angel',
 'Series J',
 'Angel Round',
 'pre-Series A',
 'Venture - Series Unknown',
 'Bridge Round',
 'Private Equity',
 'Debt and Preference capital',
 'Inhouse Funding',
 'Seed/ Angel Funding',
 'Debt',
 'Pre Series A',
 'Equity',
 'Debt-Funding',
 'Mezzanine',
 'Series B (Extension)',
 'Equity Based Funding',
 'Private Funding',
 'Seed / Angel Funding',
 'Seed/Angel Funding',
 'Seed funding',
 'Seed / Angle Funding',
 'Angel / Seed Funding',
 'Private',
 'Structured Debt',
 'Term Loan',
 'PrivateEquity',
 'Angel Funding']

In [44]:
  Mapping={
    # Pre-Series A
    'Pre-Series A': 'Pre-Series A',
    'Pre Series A': 'Pre-Series A',
    'Pre-Series A ': 'Pre-Series A',

    # Debt
    'Debt Funding': 'Debt',
    'Debt-Funding': 'Debt'}

In [45]:
df['Round']=df['Round'].replace('Mapping')

/tmp/ipykernel_57/2861844759.py:1: FutureWarning: Series.replace without 'value' and with non-dict-like 'to_replace' is deprecated and will raise in a future version. Explicitly specify the new values instead.
  df['Round']=df['Round'].replace('Mapping')


In [46]:
df['Round'].str.replace('Funding','').str.replace('Round','').str.replace('Equity','').str.strip().str.title().unique()

array(['Private', 'Series C', 'Series B', 'Pre-Series A', 'Seed',
       'Series A', 'Series D', 'Series F', 'Series E', 'Debt', 'Series G',
       'Series H', 'Venture', '', 'Corporate', 'Maiden', 'Single Venture',
       'Angel', 'Series J', 'Venture - Series Unknown', 'Bridge',
       'Debt And Preference Capital', 'Inhouse', 'Seed/ Angel',
       'Pre Series A', 'Debt-', 'Mezzanine', 'Series B (Extension)',
       'Based', 'Seed / Angel', 'Seed/Angel', 'Seed Funding',
       'Seed / Angle', 'Angel / Seed', 'Structured Debt', 'Term Loan'],
      dtype=object)

#### **Fixing the inconsistency in the 'Investors' column -**

In [47]:
df['Investors'].unique().tolist()

['Tiger Global Management',
 'Susquehanna Growth Equity',
 'Sequoia Capital India',
 'Vinod Khatumal',
 'Sprout Venture Partners',
 'Chiratae Ventures',
 'Ant Financial',
 'Sathguru Catalyzer Advisors',
 'Ping An Global Voyager Fund',
 'Mumbai Angels, Ravikanth Reddy',
 'SAIF Partners, Spring Canter Investment Ltd.',
 'Paytm, NPTK, Sabre Partners and Neoplux',
 'Vertex Growth Fund',
 'Undisclosed',
 'Ruizheng Investment',
 'Manipal Education and Medical Group (MEMG)',
 'SoftBank Vision Fund',
 'Sequoia, CapitalG, Accel',
 'Sauce.vc, Rainforest Ventures',
 'Prime Venture Partners, LetsVenture, PS1 Venture and GlobalLogic co-founder Rajul Garg',
 'RB Investments',
 'DG Daiwa Ventures, DG Incubation',
 'Trifecta Capital Advisors',
 'FinTech',
 'Dream Incubator',
 'Altimeter Capital, Sutter Hill Ventures',
 'Startup Buddy',
 'Amour Infrastructure',
 'Ackermans & van Haaren, HealthQuad, Rebright Partners, Toppan Printing',
 'Matrix Partners, Stellaris Venture Partners, Kalaari Capital',
 'V

I am using a dual-dataframe approach to handle rows with multiple Investors name 
The dataset contains multiple investor names in a single row separated by commas.  
To perform proper investor-level analysis, a separate dataframe was created.

- `df_original` is used for:
  - Funding analysis
  - Sector analysis
  - City analysis
  - Startup analysis

- `df_investor` is used for:
  - Investor analysis
  - Top investors
  - Investor participation trends



In [48]:
df['Investors']=df['Investors'].str.replace('\\\\xe2\\\\x80\\\\x99s','').str.replace('\\\\xc2\\\\xa0','').str.replace('and','')
df['Investors']=df['Investors'].str.title().str.strip()

In [49]:
df_by_investor = df.copy()
df_by_investor['Investors'] = df_by_investor['Investors'].str.split(',')
df_by_investor= df_by_investor.explode('Investors')

In [50]:
df['Investors'].str.split(',').explode()

0         Tiger Global Management
1       Susquehanna Growth Equity
2           Sequoia Capital India
3                  Vinod Khatumal
4         Sprout Venture Partners
                  ...            
2869             Ashish Dev Kapur
2870                   Rainmatter
2871                  The Hr Fund
2872    Bessemer Venture Partners
2872                Saif Partners
Name: Investors, Length: 5463, dtype: object

* The `explode()` function was applied on the investor column to separate multiple investors into individual rows for better analysis and visualization.
* The `explode()` function is used to transform list-like values in a column into separate rows.


In [51]:
df_by_investor.sample(5)

,Sr_No,Date,StartUp,Vertical,SubVertical,Location,Investors,Round,Amount
303,304,2018-04-22,Visit,Healthcare,Ai Driven Health Tech Platform,Delhi,Biz Stone,Private Equity,9.443
1546,1547,2016-07-15,Ridlr,Consumer Internet,Public Transport Mobile App,Mumbai,Innoven Capital,Private Equity,56.658
2804,2805,2015-05-27,Casa2Inns,Affordable Hotel Booking Online,Unknown,New Delhi,Kalaari Capital,Private Equity,47.215
1182,1183,2016-12-28,Bookeventz,Consumer Internet,Online Event-Booking Platform,Mumbai,Mumbai Angels,Seed Funding,9.443
1061,1062,2017-01-12,Square Yards,Consumer Internet,Property Search Aggregator Platform,Gurgaon,Andre Hoffman,Private Equity,94.430


In [52]:
#Cleaning
df_by_investor['Investors'] = df_by_investor['Investors'].str.replace(',','').str.replace(',','').str.title().str.strip()

#### **Fixing the inconsistency in the 'Location' column -**

In [53]:
df['Location'].unique()

array(['Bengaluru', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose,', 'Amritsar',
       'Delhi', 'Kormangala', 'Hyderabad', 'Burnsville', 'Menlo Park',
       'Gurugram', 'Palo Alto', 'Santa Monica', 'Singapore', 'Taramani',
       'Andheri', 'Chembur', 'Haryana', 'New York', 'Karnataka',
       'Mumbai/Bengaluru', 'Bhopal', 'Bengaluru and Gurugram',
       'India/Singapore', 'Jaipur', 'India/US', 'Nagpur', 'Indore',
       'New York, Bengaluru', 'California', 'India', 'Ahemadabad',
       'Rourkela', 'Srinagar', 'Bhubneswar', 'Chandigarh',
       'Delhi & Cambridge', 'Kolkatta', 'Kolkata', 'Coimbatore',
       'Bangalore', 'Udaipur', 'Ahemdabad', 'Bhubaneswar', 'Ahmedabad',
       'Surat', 'Goa', 'Uttar Pradesh', 'Nw Delhi', 'Gaya', 'Vadodara',
       'Trivandrum', 'Missourie', 'Panaji', 'Gwalior', 'Karur', 'Udupi',
       'Kochi', 'Agra', 'Bangalore/ Bangkok', 'Hubli', 'Kerala',
       'Kozhikode', 'US', 'Siliguri', 'USA'

In [54]:
#Clean up known messy encoding/whitespace
df['Location'].str.replace(r'\\\\xc2\\\\xa0','',regex=True).unique()

array(['Bengaluru', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose,', 'Amritsar',
       'Delhi', 'Kormangala', 'Hyderabad', 'Burnsville', 'Menlo Park',
       'Gurugram', 'Palo Alto', 'Santa Monica', 'Singapore', 'Taramani',
       'Andheri', 'Chembur', 'Haryana', 'New York', 'Karnataka',
       'Mumbai/Bengaluru', 'Bhopal', 'Bengaluru and Gurugram',
       'India/Singapore', 'Jaipur', 'India/US', 'Nagpur', 'Indore',
       'New York, Bengaluru', 'California', 'India', 'Ahemadabad',
       'Rourkela', 'Srinagar', 'Bhubneswar', 'Chandigarh',
       'Delhi & Cambridge', 'Kolkatta', 'Kolkata', 'Coimbatore',
       'Bangalore', 'Udaipur', 'Ahemdabad', 'Bhubaneswar', 'Ahmedabad',
       'Surat', 'Goa', 'Uttar Pradesh', 'Nw Delhi', 'Gaya', 'Vadodara',
       'Trivandrum', 'Missourie', 'Panaji', 'Gwalior', 'Karur', 'Udupi',
       'Kochi', 'Agra', 'Bangalore/ Bangkok', 'Hubli', 'Kerala',
       'Kozhikode', 'US', 'Siliguri', 'USA'

In [55]:
df['Location'] = df['Location'].str.replace(r'\\\\xc2\\\\xa0','',regex=True).str.strip()

In [56]:
#Standardize delimiters to a single '/'
df['Location'] = df['Location'].replace(r'\s*(/|&|and|,|&)\s*', '/', regex=True)

In [57]:
df['Location'].unique()

array(['Bengaluru', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose/', 'Amritsar',
       'Delhi', 'Kormangala', 'Hyderabad', 'Burnsville', 'Menlo Park',
       'Gurugram', 'Palo Alto', 'Santa Monica', 'Singapore', 'Taramani',
       'Andheri', 'Chembur', 'Haryana', 'New York', 'Karnataka',
       'Mumbai/Bengaluru', 'Bhopal', 'Bengaluru/Gurugram',
       'India/Singapore', 'Jaipur', 'India/US', 'Nagpur', 'Indore',
       'New York/Bengaluru', 'California', 'India', 'Ahemadabad',
       'Rourkela', 'Srinagar', 'Bhubneswar', 'Ch/igarh',
       'Delhi/Cambridge', 'Kolkatta', 'Kolkata', 'Coimbatore',
       'Bangalore', 'Udaipur', 'Ahemdabad', 'Bhubaneswar', 'Ahmedabad',
       'Surat', 'Goa', 'Uttar Pradesh', 'Nw Delhi', 'Gaya', 'Vadodara',
       'Triv/rum', 'Missourie', 'Panaji', 'Gwalior', 'Karur', 'Udupi',
       'Kochi', 'Agra', 'Bangalore/Bangkok', 'Hubli', 'Kerala',
       'Kozhikode', 'US', 'Siliguri', 'USA', 'Lucknow',

I am using a dual-dataframe approach to handle rows with multiple cities (e.g., "Mumbai / Bangalore").

**1. Why am I doing this?**
*   **To get accurate city counts:** A startup listed as "Mumbai / Bangalore" won't show up in a search for just "Mumbai." By splitting them, I can see the true activity in every city.
*   **To avoid "Money Math" errors:** Using `explode()` duplicates the row. If I'm not careful, a $10M investment will be counted twice ($10M for Mumbai + $10M for Bangalore = $20M). 

**2. The Solution:**
*   **`df` (Original):** I will use this for all **Financial Analysis** (Total/Average Funding) to ensure the dollar amounts are 100% accurate.
*   **`df_city` (Exploded):** I will use this copy for **Geographic Analysis** (City counts and maps) so that every location is represented.


In [58]:
# Create a dedicated copy for geographic analysis
df_by_city = df.copy()


In [59]:
#Split and Explode
df_by_city['Location'] = df_by_city['Location'].str.split('/')
df_by_city = df_by_city.explode('Location')
df_by_city['Location'].unique()

array(['Bengaluru', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose', '', 'Amritsar',
       'Delhi', 'Kormangala', 'Hyderabad', 'Burnsville', 'Menlo Park',
       'Gurugram', 'Palo Alto', 'Santa Monica', 'Singapore', 'Taramani',
       'Andheri', 'Chembur', 'Haryana', 'New York', 'Karnataka', 'Bhopal',
       'India', 'Jaipur', 'US', 'Nagpur', 'Indore', 'California',
       'Ahemadabad', 'Rourkela', 'Srinagar', 'Bhubneswar', 'Ch', 'igarh',
       'Cambridge', 'Kolkatta', 'Kolkata', 'Coimbatore', 'Bangalore',
       'Udaipur', 'Ahemdabad', 'Bhubaneswar', 'Ahmedabad', 'Surat', 'Goa',
       'Uttar Pradesh', 'Nw Delhi', 'Gaya', 'Vadodara', 'Triv', 'rum',
       'Missourie', 'Panaji', 'Gwalior', 'Karur', 'Udupi', 'Kochi',
       'Agra', 'Bangkok', 'Hubli', 'Kerala', 'Kozhikode', 'Siliguri',
       'USA', 'Lucknow', 'Kanpur', 'SFO', 'London', 'Seattle', 'Dubai',
       'Varanasi', 'UK', 'Jodhpur', 'Boston', 'NY', 'Belgaum',
     

**Standardize the Location name**

In [60]:
city_mapping = {
    # Spelling Fixes & Standardizing to common names
    'Bengaluru': 'Bangalore',
    'Gurugram': 'Gurgaon',
    'Kolkatta': 'Kolkata',
    'Ahemadabad': 'Ahmedabad',
    'Ahemdabad': 'Ahmedabad',
    'Nw Delhi': 'New Delhi',
    'Delhi': 'New Delhi',
    'Kormangala': 'Bangalore', # Neighborhood in Bangalore
    'Andheri': 'Mumbai',       # Neighborhood in Mumbai
    'Chembur': 'Mumbai',       # Neighborhood in Mumbai
    
    # State names mapped to their primary startup hubs 
    'Haryana': 'Gurgaon',
    'Karnataka': 'Bangalore',
    'Kerala': 'Kochi',
    'Uttar Pradesh': 'Noida',
    
    # Cleaning up common International/Combined labels
    'USA': 'US',
    'SFO': 'San Francisco',
    'NY': 'New York'
}

In [61]:
df_by_city['Location'] = df_by_city['Location'].replace(city_mapping)
df_by_city['Location'].unique()

array(['Bangalore', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose', '', 'Amritsar',
       'Hyderabad', 'Burnsville', 'Menlo Park', 'Palo Alto',
       'Santa Monica', 'Singapore', 'Taramani', 'New York', 'Bhopal',
       'India', 'Jaipur', 'US', 'Nagpur', 'Indore', 'California',
       'Ahmedabad', 'Rourkela', 'Srinagar', 'Bhubneswar', 'Ch', 'igarh',
       'Cambridge', 'Kolkata', 'Coimbatore', 'Udaipur', 'Bhubaneswar',
       'Surat', 'Goa', 'Gaya', 'Vadodara', 'Triv', 'rum', 'Missourie',
       'Panaji', 'Gwalior', 'Karur', 'Udupi', 'Kochi', 'Agra', 'Bangkok',
       'Hubli', 'Kozhikode', 'Siliguri', 'Lucknow', 'Kanpur', 'London',
       'Seattle', 'Dubai', 'Varanasi', 'UK', 'Jodhpur', 'Boston',
       'Belgaum', 'San Mateo', 'Houston', 'Global', 'Dallas'],
      dtype=object)

In [62]:
df['Location'] = df['Location'].replace(city_mapping)
df['Location'].unique()

array(['Bangalore', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose/', 'Amritsar',
       'Hyderabad', 'Burnsville', 'Menlo Park', 'Palo Alto',
       'Santa Monica', 'Singapore', 'Taramani', 'New York',
       'Mumbai/Bengaluru', 'Bhopal', 'Bengaluru/Gurugram',
       'India/Singapore', 'Jaipur', 'India/US', 'Nagpur', 'Indore',
       'New York/Bengaluru', 'California', 'India', 'Ahmedabad',
       'Rourkela', 'Srinagar', 'Bhubneswar', 'Ch/igarh',
       'Delhi/Cambridge', 'Kolkata', 'Coimbatore', 'Udaipur',
       'Bhubaneswar', 'Surat', 'Goa', 'Gaya', 'Vadodara', 'Triv/rum',
       'Missourie', 'Panaji', 'Gwalior', 'Karur', 'Udupi', 'Kochi',
       'Agra', 'Bangalore/Bangkok', 'Hubli', 'Kozhikode', 'US',
       'Siliguri', 'Lucknow', 'Kanpur', 'SFO/Bangalore', 'London',
       'Seattle/Bangalore', 'Pune/Seattle', 'Pune/Dubai', 'Bangalore/SFO',
       'Varanasi', 'New Delhi/US', 'Mumbai/UK', 'Jodhpur',
       'Hyderabad/USA

**Why Multiple DataFrames Were Created**

Separate dataframes were created to avoid incorrect aggregation during analysis.

- `df_original` → used for startup and funding analysis
- `df_by_investor` → used for investor-level analysis
- `df_by_city` → used for city-level analysis

In [63]:
# Save original cleaned dataset
df.to_csv('cleaned_startup_data.csv', index=False)

# Save investor-level dataset
df_by_investor.to_csv('startup_investor_data.csv', index=False)

# Save city-level dataset
df_by_city.to_csv('startup_city_data.csv', index=False)

**Final Cleaning Summary**

The dataset was cleaned and standardized by:
- Handling missing values
- Fixing inconsistent categorical values
- Converting funding amount to INR Crores
- Separating multi-investor rows using `explode()`
- Cleaning city and funding round information
- Preparing separate dataframes for investor-level analysis